### DISCUSSION TITLE
Proposal: replace stability relations by stability functions

### DISCUSSION THREAD

**Discussion by TedTed** [opendp/opendp#386] [OpenDP Library] [answered]
This (long-overdue) issue comes from discussions during the September 2021 OpenDP summit.

The OpenDP framework has a notion of a stability relation, and to check if a transformation satisfies a given relation, a developer has to write something like `trans.check(d_in, d_out)`. This is a significantly more complex interface than simply being able to call a function like `trans.stability(d_in)`, and get the smallest `d_out` that satisfies this relation. In particular, if you need to compute this minimal `d_out` (e.g. to debug a pipeline whose stability doesn't match what you expect), the easiest way is to write e.g. a binary search algorithm.

This also makes chaining complicated, behind the scenes: when chaining two transformation, you would have to give the framework a "hint" explaining what the intermediate value is. This would be a big usability problem, if not for the fact that OpenDP also adds backward maps and forward maps to transformations, which implement the logic above in both directions. This works, but it also means that someone writing a transformation needs to duplicate the same logic three times.

I question whether the use of stability relations and backwards maps is actually needed for any practical use case. Even in the case of complex transformations like joins, there is still a "worst-case" `d_out` for any given `d_in` — otherwise, chaining this transformation with anything is impossible anyway. So the proposal is to make stability functions (from `d_in` to `d_out`) first-class objects in the OpenDP framework, and deprecate backwards maps and stability relations. This is both a conceptual simplification and a change that would simplify the code — and simpler code is more robust and cheaper to maintain.

It's very possible that I'm wrong, and there there are actually use cases in which stability relations solve a real problem that stability functions can't solve. I've been thinking about it a while, and Salil has suggested some examples, but none of them escaped this binary classification between "there is a worst-case `d_out` that can be computed by a function" and "this is impossible to chain, even manually". In the absence of a convincing counterexample, I don't think that the proposal above has any downside.

(As a side-note, this comment & proposal isn't necessarily representing the views of my employer.)

**Comment by Shoeboxam** [comment, marked as answer]
Thanks for opening a discussion post about this. We were hashing out some ideas about this in a meeting on October 18 with Salil. The current plan (subject to change) is to simplify the general case, as you suggest, and to make the output type more complicated in any specific case of multi-parameter distances. We also want to isolate the complexity required to represent multi-parameter distances to the specific cases that need them. `make_base_gaussian` is an example of this:

```rust
// The privacy function on the base gaussian accepts a sensitivity, 
// and emits a function that relates (epsilon, delta)
GaussianMeas.privacy_function = |d_in| -> Box<dyn Fn((Q, Q))->bool> {
    return |(eps: Q, del: Q)| -> bool {
        unimplemented!()
    }
}
// ^^ Consider the output metric to be something like `SmoothedMaxDivergenceCurve`.

// A separate combinator can be used to fix epsilon, or fix delta. It emits the free parameter
// In this case we fix delta, which is passed into the constructor and moved into the privacy function closure
ApproxDPFixDelta.privacy_function = |d_in: Box<dyn Fn((Q, Q))->bool>| -> Q {
    // Can we do better here?
    return binary_search(|delta| d_in(fixed_delta, delta), bounds)
}
```
For an API very roughly like this:
```python
meas = fix_delta(make_base_gaussian(scale=1.), delta=1e-8)
epsilon = meas.privacy(d_in=sensitivity)
```

(while this post is marked up as an answer, think of it more as a response)

**Reply by TedTed** [reply]
Thanks for the very quick answer! It seems to be about *privacy* relations/functions, while this proposal is about *stability* functions. Do you have a similar plan for some stability relations? Is it a requirement that the solution for stability relations to have the same interface as privacy relations?

As a side-note, I have it in my to-do list to also summarize past discussions into a proposal on privacy functions, which is going in a similar description that what you're outlining. Is there more publicly visible documentation about this effort, by the way?

**Reply by Shoeboxam** [reply]
The primary motivation of using relations in the first place was the generality. You could say that distance curves are an "escape hatch" that allow us to simplify stability relations, and maybe also privacy relations, with the peace-of-mind that we aren't excluding future use-cases. There's certainly an appeal to have symmetry between transformations and measurements. I don't know if it is a requirement.

I think I covered the main idea above, but here's the notes from that meeting I mentioned:
https://docs.google.com/document/d/1VvEP3N4j_KDyk-mphazdPYJIUtSprkObauD9oTq7nQM/edit?usp=sharing
What medium do you think is best to document this effort? I like this discussion topic, but we could open an issue.

**Reply by TedTed** [reply]
On stability functions, I'm not sure I am convinced by the generality argument. When building software, I see it as a good practice to implement a certain piece of logic only when necessary, not in case it might be needed for some future use case that I can't precisely articulate at the moment. Adding features is generally easier than removing them, so if it appears later that I *did* need some piece of logic, it's not a big deal to simply add it then. But that's just my opinion and I won't insist if maximal generality is a strong design requirement for you =)

On privacy relations: I think that having `privacy_function` return an abstract relation (if I understand correctly what `Box<dyn Fn((Q, Q))->bool>` is) prevents you from later adding semantic information to that object, and using it to get better composition properties. For example, I think you want to be able to compose two objects returned by `GaussianMeas.privacy_function` in a way that takes into account that those are Gaussian and can use analytic and tight composition results, rather than numerical approximations from the value of the function in multiple points. I don't think that's possible if the output is an abstract relation, and this prevents you from having budget accounting logic living outside the framework.

A more generic approach would be to have a `privacy_cost` method that returns something of some type `PrivacyCost` (or some other name), and have this class implement methods for checking privacy relations, providing ε as a function of δ or vice-versa, composing multiple instances of this type, etc. Then, specific classes implementing this interface could have special cases when you compose them together. (Note: this last idea is obviously inspired by runtime type checks such as the ones you can do in Java or Python, I don't know how feasible it is to do in Rust.)

**Reply by Shoeboxam** [reply]
@TedTed This was implemented in PR #463 and released as part of 0.5, in late August. Thanks for pushing for this! I ended up adding the discrete Gaussian from CKS20, and rewrote the Gaussian in terms of the discrete Gaussian. Both of these measurements have ZeroConcentratedDivergence output measures, but you can use a combinator to cast to ε(δ) privacy curves (and another combinator to convert to fixed (ε, δ) tuples). Going forward, we can still parameterize the output measure to better support other kinds of composition, should we need it.

**Reply by TedTed** [reply]
Thanks for the update!
